File: src/org/irislang/jiris/core/IrisObject.java
package org.irislang.jiris.core;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Map.Entry;

import org.irislang.jiris.core.IrisMethod.CallSide;
import org.irislang.jiris.core.IrisMethod.MethodAuthority;
import org.irislang.jiris.dev.IrisDevUtil;

public class IrisObject implements IrisRunningObject {
	private IrisClass m_class = null;
	private HashMap<String, IrisMethod> m_methods = new HashMap<String, IrisMethod>();
	private HashMap<String, IrisValue> m_instanceValues = new HashMap<String, IrisValue>();
	private Object m_nativeObject = null;
	private static int sm_objectNumber = 0;
	private int m_objectID = 0;
	
	public IrisObject() {
		m_objectID = ++sm_objectNumber;
	}
	
	public int getObjectID() {
		return m_objectID;
	}
	
	public IrisValue CallInstanceMethod(String methodName, ArrayList<IrisValue> parameterList, IrisContextEnvironment context, IrisThreadInfo threadInfo, IrisMethod.CallSide callSide) throws Throwable {
		IrisMethod method = null;
		boolean isCurrentMethod = false;
		
		// Object's instance method
		if(m_methods.containsKey(methodName)) {
			method = m_methods.get(methodName);
			isCurrentMethod = true;
		// instance method in class
		} else {
 			IrisClass.SearchResult result = new IrisClass.SearchResult();
			m_class.GetMethod(methodName, result);
 			method = result.getMethod();
			isCurrentMethod = result.isCurrentClassMethod();
		}
		
		if(method == null) {
			/* Error */
			//return CallInstanceMethod("missing_method", parameterList, context, callSide);
			return IrisDevUtil.Nil();
		}
		
		// Inside call
		IrisValue callResult = null;
		IrisValue caller = IrisValue.WrapObject(this);
		if(callSide == CallSide.Inside) {
			
			if(isCurrentMethod) {
				callResult = method.Call(caller, parameterList, context, threadInfo);
			}
			else {
				if(method.getAuthority() == MethodAuthority.Personal) {
					/* Error */
					callResult = IrisDevUtil.Nil();
				}
				else {
					callResult = method.Call(caller, parameterList, context, threadInfo);
				}
			}
		}
		// Outside call
		else {
			if(method.getAuthority() != MethodAuthority.Everyone) {
				/* Error */
				callResult = IrisDevUtil.Nil();
			}
			else {
				callResult = method.Call(caller, parameterList, context, threadInfo);
			}
		}
		
		return callResult;
	}

	public void AddInstanceMethod(IrisMethod method) {
		m_methods.put(method.getMethodName(), method);
	}
	
	public void AddInstanceVariable(String name, IrisValue value) {
		m_instanceValues.put(name, value);
	}
	
	public IrisValue GetInstanceVariable(String name) {
		return m_instanceValues.get(name);
	}
	
	public IrisClass getObjectClass() {
		return m_class;
	}

	public void setObjectClass(IrisClass classObj) {
		m_class = classObj;
	}

	public Object getNativeObject() {
		return m_nativeObject;
	}

	public void setNativeObject(Object nativeObject) {
		m_nativeObject = nativeObject;
	}
	
	public void ResetAllMethodsObject() throws Throwable {
		Iterator<?> iterator = m_methods.entrySet().iterator();
		while(iterator.hasNext()) {
			@SuppressWarnings("unchecked")
			Map.Entry<String, IrisMethod> entry = (Entry<String, IrisMethod>) iterator.next();
			IrisMethod method = entry.getValue();
			method.ResetMethodObject();
		}
	}

	
}
 

File: src/main/java/org/irislang/jiris/core/IrisObject.kt
package org.irislang.jiris.core

import java.util.ArrayList
import java.util.HashMap
import kotlin.collections.Map.Entry

import org.irislang.jiris.core.IrisMethod.CallSide
import org.irislang.jiris.core.IrisMethod.MethodAuthority
import org.irislang.jiris.core.exceptions.IrisExceptionBase
import org.irislang.jiris.core.exceptions.fatal.IrisInvalidAuthorityException
import org.irislang.jiris.dev.IrisDevUtil

import javax.swing.undo.AbstractUndoableEdit

class IrisObject : IrisRunningObject {
    var objectClass: IrisClass? = null
    private val m_methods = HashMap<String, IrisMethod>()
    private val m_instanceValues = HashMap<String, IrisValue>()
    var nativeObject: Any? = null
    var objectID = 0

    init {
        objectID = ++sm_objectNumber
    }

    @Throws(IrisExceptionBase::class)
    fun CallInstanceMethod(methodName: String, parameterList: ArrayList<IrisValue>,
                           context: IrisContextEnvironment, threadInfo: IrisThreadInfo, callSide: IrisMethod.CallSide): IrisValue? {
        var method: IrisMethod? = null
        var isCurrentMethod = false

        var authority: MethodAuthority? = null

        // Object's instance method
        if (m_methods.containsKey(methodName)) {
            method = m_methods[methodName]
            isCurrentMethod = true
            authority = method!!.authority
            // instance method in class
        } else {
            val result = IrisClass.SearchResult()
            objectClass!!.GetMethod(methodName, result)
            method = result.method

            if (method == null) {
                /* Error */
                val mName = IrisDevUtil.CreateString(methodName)
                val mClsName = IrisDevUtil.CreateString(objectClass!!.className)
                val tmpList = ArrayList<IrisValue>(2)
                tmpList.add(mName)
                tmpList.add(mClsName)
                return IrisDevUtil.CallMethod(IrisValue.WrapObject(this), "missing_method", tmpList, context, threadInfo)
            }

            isCurrentMethod = result.isCurrentClassMethod

            if (result.isCurrentClassMethodOfSelf) {
                authority = method.authority
            } else {
                authority = objectClass!!.GetMethodAuthorityFromMap(methodName)
                if (authority == null) {
                    authority = method.authority
                }
            }
        }

        // Inside call
        var callResult: IrisValue? = null
        val caller = IrisValue.WrapObject(this)
        if (callSide == CallSide.Inside) {

            if (isCurrentMethod) {
                callResult = method.Call(caller, parameterList, context, threadInfo)
            } else {
                if (authority == MethodAuthority.Personal) {
                    /* Error */
                    throw IrisInvalidAuthorityException(threadInfo.currentFileName, threadInfo
                            .currentLineNumber, "Method with personal authority cannot be called here.")
                } else {
                    callResult = method.Call(caller, parameterList, context, threadInfo)
                }
            }
        } else {
            if (authority != MethodAuthority.Everyone) {
                /* Error */
                throw IrisInvalidAuthorityException(threadInfo.currentFileName, threadInfo
                        .currentLineNumber, "Only method with everyone authority can be called here.")
            } else {
                callResult = method.Call(caller, parameterList, context, threadInfo)
            }
        }// Outside call

        return callResult
    }

    fun AddInstanceMethod(method: IrisMethod) {
        m_methods[method.methodName] = method
    }

    fun AddInstanceVariable(name: String, value: IrisValue) {
        m_instanceValues[name] = value
    }

    fun GetInstanceVariable(name: String): IrisValue? {
        return m_instanceValues[name]
    }

    @Throws(IrisExceptionBase::class)
    fun ResetAllMethodsObject() {
//        val iterator = m_methods.entries.iterator()
//        while (iterator.hasNext()) {
//            val entry = iterator.next() as Entry<String, IrisMethod>
//            val method = entry.value
//            method.ResetMethodObject()
//        }

        for (iter in m_methods) {
            iter.value.ResetMethodObject()
        }
    }

    internal fun GetInstanceMethod(methodName: String): IrisMethod? {
        return if (m_methods.containsKey(methodName)) {
            m_methods[methodName]
        } else {
            null
        }
    }

    companion object {
        private var sm_objectNumber = 0
    }

}
 